daml-script/runner/src/main/scala/com/digitalasset/daml/lf/Free.scala

```diff
@@ -0,0 +1,289 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package engine
+package free
+
+import data.Ref
+import speedy._
+import SExpr.SExpr
+import com.daml.logging.LoggingContext
+import scalaz.std.either._
+import scalaz.std.vector._
+import scalaz.syntax.traverse._
+
+private[lf] sealed abstract class Result[+X, +Q] extends Product with Serializable {
+  def transform[Y, R >: Q](f: Result.ErrOr[X] => Result[Y, R]): Result[Y, R]
+  def flatMap[Y, R >: Q](f: X => Result[Y, R]): Result[Y, R] =
+    transform((err: Result.ErrOr[X]) => err.fold(Result.failed, f))
+  def map[Y, R >: Q](f: X => Y): Result[Y, R] = flatMap(x => Result.successful(f(x)))
+  def remapQ[R](f: Q => Result[Result.ErrOr[SExpr], R]): Result[X, R]
+}
+
+private[lf] object Result {
+
+  type ErrOr[+X] = Either[RuntimeException, X]
+
+  object Implicits {
+    final implicit class ErrOrOps[X](val either: ErrOr[X]) extends AnyVal {
+      def toResult: Final[X] = Final(either)
+    }
+  }
+
+  final case class Final[X](x: ErrOr[X]) extends Result[X, Nothing] {
+    override def transform[Y, Q](f: ErrOr[X] => Result[Y, Q]): Result[Y, Q] = f(x)
+    override def remapQ[R](f: Nothing => Result[Result.ErrOr[SExpr], R]): Result[X, Nothing] = this
+  }
+
+  def successful[X](x: X): Final[X] = Final(Right(x))
+
+  def failed(err: RuntimeException): Final[Nothing] = Final(Left(err))
+
+  final case class Interruption[X, Q](resume: () => Result[X, Q]) extends Result[X, Q] {
+    override def transform[Y, R >: Q](f: ErrOr[X] => Result[Y, R]): Result[Y, R] =
+      copy(() => resume().transform(f))
+    override def remapQ[R](f: Q => Result[ErrOr[SExpr], R]): Result[X, R] =
+      Interruption(() => resume().remapQ(f))
+  }
+
+  final case class Question[X, Q](
+      q: Q,
+      resume: ErrOr[SExpr] => Result[X, Q],
+  ) extends Result[X, Q] {
+
+    override def transform[Y, R >: Q](f: ErrOr[X] => Result[Y, R]): Result[Y, R] =
+      copy(resume = resume(_).transform(f))
+
+    override def remapQ[R](f: Q => Result[ErrOr[SExpr], R]): Result[X, R] =
+      f(q).flatMap(resume(_).remapQ(f))
+  }
+
+  val Unit = Result.Final(Right(()))
+}
+
+case class ConversionError(message: String) extends RuntimeException(message)
```

Comment:
Surely, we should have this as final and private? Particularly worried that outside users might try and compare multiple instances on this as we are in an abstract class.